### ISSUE TITLE
Incorrect type inferred for and-or expression

### ISSUE THREAD

**Issue by nludban** [microsoft/pyright#7112]
**Describe the bug**

Incorrect type inferred for and-or expression with constant false value (eg, `None`), the result is never a `str`.

`error: Expression of type "int | Literal[''] | None" cannot be assigned to return type "int | None" (reportGeneralTypeIssues)`

**Code or Screenshots**

```python
def simple(self, s: str) -> int | None:
    return (s or None) and int(s)
```

`return int(s) if s else None` is a workaround.

**VS Code extension or command-line**

Command line: pyright 1.1.345

**Comment by erictraut** [comment]
Pyright is working as designed here. I don't consider this an error.

Expression types are evaluated by a type checker according to order of operations — the same order that they are evaluated at runtime by the interpreter. That means pyright first evaluates the type of the expression `s or None`. Its type is `str | None`, since it cannot be narrowed further. 

You'll need to change the code if you want it to type check in pyright.
